[SPARK-26735][SQL] Verify plan integrity for special expressions

rednaxelafx opened the pull request:
## What changes were proposed in this pull request?

Add verification of plan integrity with regards to special expressions being hosted only in supported operators. Specifically:

- `AggregateExpression`: should only be hosted in `Aggregate`, or indirectly in `Window`
- `WindowExpression`: should only be hosted in `Window`
- `Generator`: should only be hosted in `Generate`

This will help us catch errors in future optimizer rules that incorrectly hoist special expression out of their supported operator.

TODO: This PR actually caught a bug in the analyzer in the test case `SPARK-23957 Remove redundant sort from subquery plan(scalar subquery)` in `SubquerySuite`, where a `max()` aggregate function is hosted in a `Sort` operator in the analyzed plan, which is invalid. That test case is disabled in this PR.
SPARK-26741 has been opened to track the fix in the analyzer.

## How was this patch tested?

Added new test case in `OptimizerStructuralIntegrityCheckerSuite`

Commit:
verify plan integrity: special expressions AggregateExpression, WindowExpression and Generator should only be hosted in corresponding operators

SparkQA commented:
**[Test build #101699 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101699/testReport)** for PR 23658 at commit [`cf3f288`](https://github.com/apache/spark/commit/cf3f288de8e440310eefc99ebb3916e591e6f62d).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7440/
Test PASSed.

cloud-fan commented:
I like this check! It can save us a lot of time when debugging!

cloud-fan commented:
LGTM except a code style comment.

SparkQA commented:
**[Test build #101699 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101699/testReport)** for PR 23658 at commit [`cf3f288`](https://github.com/apache/spark/commit/cf3f288de8e440310eefc99ebb3916e591e6f62d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101699/
Test FAILed.

Commit:
Address comments

SparkQA commented:
**[Test build #101712 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101712/testReport)** for PR 23658 at commit [`c7ccc42`](https://github.com/apache/spark/commit/c7ccc42c18d46d638c42c4fe13f8494393908f2f).

rednaxelafx commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -41,7 +41,39 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
   // Check for structural integrity of the plan in test mode. Currently we only check if a plan is
   // still resolved after the execution of each rule.
```
Thanks! Updated

rednaxelafx commented:
Thanks for your reviews, @cloud-fan and @viirya !

I've updated the PR addressing your comments.

This PR has actually caught a genuine bug in the analyzer in one of the test cases:
For this query in `SubquerySuite`,
```sql
SELECT *
FROM   t1
WHERE  c1 = (SELECT   max(t2.c1)
             FROM     t2
             GROUP BY t2.c1
             HAVING   count(*) >= 1
             ORDER BY max(t2.c1))
```
The analyzer resolves it into:
```
== Analyzed Logical Plan ==
c1: int, c2: int
Project [c1#51, c2#52]
+- Filter (c1#51 = scalar-subquery#109 [])
   :  +- Project [max(c1)#137]
   :     +- Sort [max(c1#28) ASC NULLS FIRST], true                // NOTE HERE!!
   :        +- Project [max(c1)#137, c1#28]
   :           +- Filter (count(1)#139L >= cast(1 as bigint))
   :              +- Aggregate [c1#28], [max(c1#28) AS max(c1)#137, count(1) AS count(1)#139L, c1#28]
   :                 +- SubqueryAlias `t2`
   :                    +- Project [_1#23 AS c1#28, _2#24 AS c2#29]
   :                       +- LocalRelation [_1#23, _2#24]
   +- SubqueryAlias `t1`
      +- Project [_1#46 AS c1#51, _2#47 AS c2#52]
         +- LocalRelation [_1#46, _2#47]
```
... where `Sort [max(c1#28) ASC NULLS FIRST], true` is an example of a `Sort` operator hosting an aggregate expression `max`, incorrectly.

It's somewhat tedious to fix because we need to tweak the order a bit. Working on it.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7449/
Test PASSed.

SparkQA commented:
**[Test build #101712 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101712/testReport)** for PR 23658 at commit [`c7ccc42`](https://github.com/apache/spark/commit/c7ccc42c18d46d638c42c4fe13f8494393908f2f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101712/
Test FAILed.

Commit:
Disable failing test case in SubquerySuite. It should be re-enabled after fixing SPARK-26741

SparkQA commented:
**[Test build #101718 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101718/testReport)** for PR 23658 at commit [`4c34e98`](https://github.com/apache/spark/commit/4c34e98973b63323013f65d652fd7c14ab00c0fb).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7454/
Test PASSed.

SparkQA commented:
**[Test build #101718 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101718/testReport)** for PR 23658 at commit [`4c34e98`](https://github.com/apache/spark/commit/4c34e98973b63323013f65d652fd7c14ab00c0fb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101718/
Test FAILed.

Commit:
Address Xiao's comment and fix test

SparkQA commented:
**[Test build #101721 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101721/testReport)** for PR 23658 at commit [`c1d38dd`](https://github.com/apache/spark/commit/c1d38dd2fc05237131baaac443a493593c2fa863).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7456/
Test PASSed.

SparkQA commented:
**[Test build #101721 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101721/testReport)** for PR 23658 at commit [`c1d38dd`](https://github.com/apache/spark/commit/c1d38dd2fc05237131baaac443a493593c2fa863).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101721/
Test PASSed.
